lib/src/embedded/isolate_dispatcher.dart

```diff
@@ -102,84 +106,67 @@ class IsolateDispatcher {
     }, onError: (Object error, StackTrace stackTrace) {
       _handleError(error, stackTrace);
     }, onDone: () async {
-      _closed = true;
-      for (var isolate in _allIsolates) {
+      for (var isolate in _isolates) {
         (await isolate).kill();
       }
 
-      // Killing isolates isn't sufficient to make sure the process closes; we
-      // also have to close all the [ReceivePort]s we've constructed (by closing
-      // the [IsolateChannel]s).
-      for (var sink in _activeIsolates.values) {
+      for (var sink in _sinks) {
         sink.close();
       }
-      for (var channel in _inactiveIsolates) {
-        channel.sink.close();
-      }
     });
   }
 
-  /// Returns an isolate that's ready to run a new compilation.
+  /// Returns the mailbox for an isolate that's ready to run a new compilation.
   ///
   /// This re-uses an existing isolate if possible, and spawns a new one
   /// otherwise.
-  Future<StreamSink<Uint8List>> _getIsolate(int compilationId) async {
+  Future<Mailbox> _getMailbox(int compilationId) async {
     var resource = await _isolatePool.request();
-    if (_inactiveIsolates.isNotEmpty) {
-      return _activate(_inactiveIsolates.first, compilationId, resource);
+    if (_inactive.isNotEmpty) {
+      var lease = _inactive.first;
+      _inactive.remove(lease);
+      lease.id = compilationId;
+      lease.resource = resource;
+      _mailboxes[compilationId] = lease.mailbox;
+      return lease.mailbox;
     }
 
+    var mailbox = Mailbox();
+    var lease = _Lease(mailbox, compilationId, resource);
+    _mailboxes[compilationId] = mailbox;
+
     var receivePort = ReceivePort();
-    var future = Isolate.spawn(_isolateMain, receivePort.sendPort);
-    _allIsolates.add(future);
+    var future =
+        Isolate.spawn(_isolateMain, (mailbox.asSendable, receivePort.sendPort));
+    _isolates.add(future);
     await future;
 
-    var channel = IsolateChannel<_InitialMessage?>.connectReceive(receivePort)
+    var channel = IsolateChannel<Uint8List?>.connectReceive(receivePort)
```

Comment:
Do we still need a two-way channel here? Could we just use the `Mailbox` to handle all the message sends, including those that start new compilations?

Reply:
Unfortunately we cannot use mailbox. Mailbox currently only support one message at a time therefore we cannot have multiple isolate sending message to a single mailbox.

Reply:
The reason I kept `IsolateChannel` to have two way connection is that `ExplicitCloseTransformer` was super handy for cleaning up the sink on either side, so that cleaning up on crash is easier. This can definitely be just a one-way stream.